### PR TITLE
feat(ness): migrate harness safety guard from Python to Rust

### PR DESCRIPTION
**Pull request by phatblat**
## Summary
Every coding-agent harness (Claude, Codex, Cursor, Grok, Crush, Antigravity, OpenCode, Pi, omp) used a duplicated Python script to block dangerous commands, protected-file writes, and secret-like content before a tool call ran. This replaces all of those per-harness Python guards with a single compiled Rust binary (`ness`), so every harness now shares one tested policy instead of drifting copies.

## Changes
- Cut every harness hook (Claude, Codex, Cursor, Grok, Crush, Antigravity, OpenCode, Pi, omp) over to the compiled `ness` guard and delete the legacy Python guard (`safety.py`) and per-adapter `harness-guard.py` scripts
- Wire the compiled guard into CI parity checks and `cargo install`
- Add an `install-rust-deps` just recipe for the Rust toolchain and llvm-tools
- Update stale `just` recipe references and residual `safety.py` references after the migration
- Gate the required lint check on `hk.pkl` changes
- Resync `docs/harness/observed.json` guard-surface evidence with the new `.sh` adapter shims

### CODE REVIEW

File: crates/ness/tests/corpus.rs [hunk 1]
```diff
@@ -0,0 +1,458 @@
+//! Corpus and shim conformance tests for `ness`, replacing
+//! `scripts/ness-parity.py` and the guard-behavior bats tests it exercised.
```
Comment (phatblat) [inline crates/ness/tests/corpus.rs:2]: **Finding #4:** Coverage gap vs. the deleted bats suite: the removed `tests/agent-harnesses.bats` test "guard allows reads and unrelated writes near the control plane" asserted that read-only commands merely *naming* a control-plane path (`cat safety.py`, `grep -n deny agent-harnesses.py`) are still allowed — guarding against an overzealous "command string contains the path" check. No corpus fixture reproduces this scenario post-migration. `evaluate_command`/`candidate_write_targets` are unchanged by this PR so there's no functional regression, but this specific edge case has no regression test anymore. Worth adding a `corpus/` case for it.

File: crates/ness/tests/corpus/090-deny-controlplane-hooks-dir.json [hunk 2]
```diff
@@ -2,6 +2,10 @@
   "command": null,
   "content": "x",
   "cwd": null,
+  "expected": {
+    "decision": "deny",
+    "reason": "control-plane file is human-only: /Users/phatblat/.agents/harness/hooks/safety.py. It decides what the guard permits, so edit it directly rather than through an agent."
+  },
   "harness": "claude",
   "path": "/Users/phatblat/.agents/harness/hooks/safety.py",
```
Comment (phatblat) [inline crates/ness/tests/corpus/090-deny-controlplane-hooks-dir.json]: **Finding #5:** Cosmetic: this fixture (and `100-deny-controlplane-via-bash-command-substring-not-file.json`) uses `.agents/harness/hooks/safety.py` as its example path under the protected `hooks/` directory, but that file no longer exists after this PR. Harmless — the check is a pure string-pattern match, not filesystem-based — but a made-up filename (or `contract.json`, which does still exist there) would read less confusingly.

File: crates/ness/src/hook.rs [hunk 3]
```diff
@@ -1,17 +1,26 @@
...
+//! sniffing `tool_name`, which differs across harnesses (Claude's Bash tool
+//! is not necessarily named the same as Codex's).
+//!
+//! `grok`, `crush`, `antigravity`, and `cursor` take no `--tool`: their
+//! payloads carry a single normalized tool call (`toolInput`/`tool_input`
+//! for grok/crush, or top-level `tool`/`command`/`path`/`content` for
+//! antigravity/cursor) that `classify` sorts into bash-vs-write the same way
+//! the Python wrappers did (`render_grok_guard`, `render_crush_guard`,
+//! `render_antigravity_guard`/`render_cursor_guard`,
+//! `scripts/agent-harnesses.py`).
```
Comment (phatblat) [inline crates/ness/src/hook.rs:20]: **Finding #6:** Pre-existing, out-of-scope observation for a follow-up: this doc comment highlights that grok/crush/antigravity/cursor now fail *closed* on an unreadable stdin payload via `read_stdin_object()`, but the Claude/Codex path above (`read_stdin()`) still fails *open* to `Value::Null` on malformed JSON (an empty command/path is then just treated as "nothing to check" → allow). Since this PR is explicitly extending the fail-closed guarantee to four more harnesses, closing the same gap for Claude/Codex would make the guarantee uniform across all six.

**Comment by phatblat** [comment]
**Finding #1** (docs/agent-session-transcripts.md:16): Stale reference: `.agents/harness/hooks/safety.py:71` no longer exists — this PR deletes `safety.py` and moves the policy to `crates/ness/src/policy.rs` (`PROTECTED_PATHS`). Worth a follow-up doc pass so this pointer doesn't dangle.

_Originally at: https://github.com/phatblat/dotfiles/blob/82c10b02e09e611e636ff02db8e5af87950664a8/docs/agent-session-transcripts.md#L16_

**Comment by phatblat** [comment]
**Finding #2** (docs/specs/001-self-improvement-loop.md:24): Stale reference: this spec's `sources` list and several inline mentions (also around L57-62, L86-89, L141-142) still cite `.agents/harness/hooks/safety.py`, which this PR deletes in favor of `crates/ness/src/policy.rs`. Not blocking — this file isn't touched by this PR — but the plan doc is now stale.

_Originally at: https://github.com/phatblat/dotfiles/blob/82c10b02e09e611e636ff02db8e5af87950664a8/docs/specs/001-self-improvement-loop.md#L24_

**Comment by phatblat** [comment]
**Finding #3** (.typos.toml:18): Stale reference: comment cites `.agents/harness/hooks/safety.py`, which this PR deletes. The regex exclusion itself is still correct (it also matches the Rust port in `crates/ness`), but the comment's file pointer is now wrong.

_Originally at: https://github.com/phatblat/dotfiles/blob/82c10b02e09e611e636ff02db8e5af87950664a8/.typos.toml#L18_

**Comment by phatblat** [comment]
Resolved by d413 — docs/agent-session-transcripts.md:16 repointed the stale safety.py reference to crates/ness/src/policy.rs.

**Comment by phatblat** [comment]
Resolved by fdf1 — .typos.toml:18's comment repointed to crates/ness/src/policy.rs; the exclusion pattern itself was already correct.

**Comment by phatblat** [comment]
Not applying — docs/specs/001-self-improvement-loop.md is a `status: approved`, `approved_at: 2026-08-12` decision record documenting evidence and a verification transcript **as they existed at approval time**, before this PR's Python→Rust migration existed. The `sources` list and the `guard --tool write --path .agents/harness/hooks/safety.py → allow` transcript are historical facts about what was run and approved that day, not a living reference to the current guard implementation. Rewriting them to cite crates/ness/src/policy.rs would misrepresent what was actually verified when the spec was approved. Leaving this open for a maintainer call rather than resolving it myself.

File: crates/ness/tests/corpus.rs (hunk 1 above)
Comment (phatblat) [reply crates/ness/tests/corpus.rs:2]: Resolved by d7bdf — added 103-allow-controlplane-bash-read-only.json covering chained cat+grep over control-plane paths, expected to allow.

File: crates/ness/tests/corpus/090-deny-controlplane-hooks-dir.json (hunk 2 above)
Comment (phatblat) [reply crates/ness/tests/corpus/090-deny-controlplane-hooks-dir.json]: Resolved by aa53 — fixtures 090 and 100 now use contract.json (an existing protected file) instead of the deleted safety.py.

File: crates/ness/src/hook.rs (hunk 3 above)
Comment (phatblat) [reply crates/ness/src/hook.rs:20]: Acknowledged as a follow-up — you flagged this yourself as "pre-existing, out-of-scope" for this PR. Not making the Claude/Codex read_stdin() path fail-closed here since that's an unrelated behavior change beyond this PR's Python→Rust guard cutover; tracking as a candidate follow-up rather than fixing inline.
